### PR TITLE
feat(packages/excalidraw): Extended fonts support

### PR DESCRIPTION
**Pull request by tothatt81**
## Summary

Adds provider-backed custom fonts. Hosts expose a font catalog and lazily resolve font definitions, so no font has to be bundled with Excalidraw. Custom families work across the font picker, scene persistence, loading, paste and duplication, canvas/SVG exports, and the headless export utilities.

Custom font ids are persisted as `<provider>:<family>`, e.g. `google:Roboto`.

## API

Provider keys are lowercase, `:`-free, and **persisted** — they travel with the document across collab, export and reload, so a key must always denote the same font source. Family names are matched exactly; a resolver should reject anything it does not recognise verbatim rather than normalising it.

Every definition must carry at least one fetchable, non-`local()` font face. Metadata-only and system/local definitions are rejected, so a custom font is always loadable and inlineable into SVG exports.

Headless exports take the matching resolvers:

```ts
await exportToSvg({ elements, appState, files, fontResolvers });
```

`Fonts.registerCustomFont()` is available for hosts that already hold a definition and want to register it eagerly.

## Breaking changes

`fontFamily` is now `number | "<provider>:<family>"` instead of a numeric union. This affects `ExcalidrawTextElement["fontFamily"]`, `AppState.currentItemFontFamily` and `AppState.currentHoveredFontFamily`. Built-in fonts keep their numeric ids and existing scenes are unaffected.

## Behavior

- Provider fonts appear in the font picker, resolved lazily as rows become visible rather than loading the whole catalog.
- Fonts outside `availableFonts` can be added by exact-name search, with loading, failure and explicit retry states.
- Provider-qualified families are persisted in elements and app state, and restored without being coerced to a built-in font.
- Custom fonts are loaded for pasted, duplicated, restored and externally updated elements, and embedded into SVG exports with their weight, style and unicode-range descriptors preserved.
- Concurrent resolutions are deduplicated and capped; browser font-file failures are kept distinct from provider-resolution failures.
- Asynchronous text repairs are deferred while a gesture is in flight, so they cannot corrupt the gesture's undo entry.
- Without `fontProviders`, the feature costs nothing: no observers, no per-update element scans, no extra font work.

## Design notes

- **Provider ids are persisted identifiers.** A provider id must always refer to the same font source, and a qualified family must resolve to an equivalent definition for the lifetime of the page.
- **Resolution is page-global, scene processing is per instance.** Browser font faces are document-level resources, so resolved definitions and failures are shared by every editor on the page; each editor still processes its own scene.
- **Geometry is reconciled once per family, per editor**, when that family loads. Elements introduced afterwards keep their persisted or host-supplied geometry — including duplicates whose `fontFamily` a host swaps in `onDuplicate`.
- **Exports render persisted geometry as-is.** They resolve and embed fonts, but do not re-measure text that was persisted with fallback measurements.
- **Resolver concurrency is best-effort.** A timed-out promise cannot be cancelled and may keep running after its queue slot is released.
- The picker's Scene Fonts group refreshes on selection and font-registration changes rather than on every scene mutation while the picker is open.

## Testing

- Provider-family parsing, restoration and app-state normalization.
- Registry and resolver behavior: validation, failure, retry, dedup, concurrency, timeout, cross-instance sharing, and the no-providers zero-cost path.
- Font picker: catalog, lazy visibility-driven registration, search, retry, keyboard navigation, stale-resolution cancellation and lifecycle.
- Paste, duplication, bound-text reflow, and the deferred-repair/undo-history regression.
- Font-face descriptor emission and SVG `@font-face` declarations; affected snapshots updated.

### CODE REVIEW

**Comment by dwelle** [comment]
hi - quick fable review before I can take a look:

-----

## Bugs

**1\. SVG export is broken for custom fonts** — [ExcalidrawFontFace.ts:50](/packages/excalidraw/fonts/ExcalidrawFontFace.ts#L50) emits `@font-face { font-family: ${this.fontFace.family}; ... }` unquoted. For `google:Roboto` that's invalid CSS (colons can't appear in an identifier), so the whole `@font-face` rule is dropped and the exported SVG renders the fallback font, while the text elements reference the quoted `"google:Roboto"` family. Fix: quote and escape the family in `toCSS()` (safe for all built-ins too, same escaping you added in `getFontFamilyString`).

**2\. Failed resolutions retry forever** — in [App.tsx:3650-3657](/packages/excalidraw/components/App.tsx#L3650-L3657), families that fail to resolve are deleted from `customFontsAttempted`, so the next `componentDidUpdate` (which fires constantly) retries them. A scene containing one permanently-unresolvable font (typo, font missing from your provider, network down) turns into an endless stream of provider requests and `console.error` spam. This also contradicts the field's own doc comment ("Prevents re-triggering … including failed resolutions"). Keep failures in the set, or add a bounded retry/backoff.

**3\. Registered-but-never-loaded custom fonts are skipped by the loaders** — in both `loadSceneFonts` and `loadElementsFonts` ([Fonts.ts:216-242](/packages/excalidraw/fonts/Fonts.ts#L216-L242)), `customFamilies` filters out strings already in `Fonts.registered`, and string families are also excluded from `builtInFamilies` — so a registered string family gets **no** `document.fonts.load()` call at all. Registration ≠ loaded: the picker's open-effect and the public `Fonts.registerCustomFont` API both register fonts without loading them (relying on CSS lazy-load, which canvas `fillText` never triggers). Repro: open the picker with a search that filters font X out of the DOM (registered, never rendered), then paste elements using X — `loadElementsFonts` skips it, the auto-detect in App skips it too (`Fonts.registered.has(f)`), and the canvas shows the fallback with nothing to re-trigger a load. The `!Fonts.registered.has(f)` filter belongs to the *resolution* step only (and `resolveCustomFonts` already filters internally); the *load* step should get all string families.

**4\. Restore doesn't do what the PR description claims** — "non-prefixed unknown font names from restore/import are normalized back to default" only holds for the legacy `element.font` path via `getFontFamilyByName`. A modern file with `fontFamily: "Random Font"` passes through [restore.ts:516](/packages/excalidraw/data/restore.ts#L516) untouched, becomes a "custom" family, and renders whatever local font CSS resolves. Either validate `fontFamily` itself in restore (keep numeric + provider-prefixed, default the rest) or amend the PR description. Related compat note: string `fontFamily` now flows to collab peers and older clients; they degrade to fallback fonts, which is probably acceptable but worth stating in the PR.

## Room for improvement

- **Eager resolve of the whole catalog**: the picker-open effect in [FontPickerList.tsx:411](/packages/excalidraw/components/FontPicker/FontPickerList.tsx#L411) calls `provider.resolve()` for *every* `availableFonts` entry — N network round-trips on first open. Fine for a curated dozen; bad for a Google-Fonts-sized catalog. Consider resolving only items as they render, or letting providers supply lightweight previews.
- **Dead dedupe**: in `allFonts`, `builtInValues` contains numbers while provider values are strings, so `providerFonts.filter((f) => !builtInValues.has(f.value))` never filters anything, and the "already registered get skipped" comment doesn't match any check. A provider exposing e.g. "Nunito" will produce a duplicate row next to the built-in one — dedupe by display name if that's the intent.
- **Per-update scan cost**: with providers configured, every `componentDidUpdate` rebuilds the scene-family set (O(elements)), and a hit calls `loadSceneFonts()` which re-processes built-ins too. Consider gating on scene version/elements actually changing, and loading just the new families.
- **Triplicated resolve logic**: `Fonts.resolveCustomFonts`, the picker-open effect, and `handleUseCustomFont` each reimplement parse → resolve → register (and the built-in/custom split is copy-pasted between `loadSceneFonts` and `loadElementsFonts`). Extracting one helper would shrink the diff meaningfully and keep the flows from drifting.
- **Cross-browser check for the family name**: `new FontFace("google:Roboto", …)` relies on browsers accepting arbitrary family strings. Modern Chrome/Firefox/Safari do, but older Safari/Firefox parsed the constructor argument as a CSS `<family-name>` and threw on invalid names. Worth one manual check in Safari before merging.
- **Provider API friction**: `FontDefinition.metrics` requires exact `unitsPerEm`/`ascender`/`descender`, which providers generally can't supply without parsing the font binary. Consider making metrics optional with sensible defaults, since `getLineHeight` already has a fallback. Also, `handleUseCustomFont` with a search term that itself contains a prefix ("google:Roboto") produces `google:google:Roboto` — worth parsing the term first.
- **UX nits**: the custom-font row shows only the raw term — a label like "Use font "X"" (i18n'd) would clarify that clicking triggers a fetch; failures only `console.warn` with no user-visible feedback; and Enter can fire a provider fetch in the brief window before the first filtered result gets auto-hovered.
- **No tests**: the PR adds parsing (`parseProviderFontFamily` edge cases: `":x"`, `"x:"`, `"a:b:c"`), restore normalization, registry behavior, and picker flows — all easily unit-testable, and none covered. Given the repo's `yarn test:update` convention, at least the pure helpers and restore behavior deserve tests.

File: packages/excalidraw/components/App.tsx
Comment (mtolmacs) [inline packages/excalidraw/components/App.tsx:1]: Move the logic out of `App.tsx` into `App.font.tsx` or similar. See other examples already on master. The goal is to use the App class as a dispatcher and the workflow/feature specific logic lives in a helper class (i.e. `class AppFonts {}`.

File: packages/common/src/font-metadata.ts [hunk 1]
```diff
@@ -157,9 +200,9 @@ export const getVerticalOffset = (
   fontSize: ExcalidrawTextElement["fontSize"],
   lineHeightPx: number,
 ) => {
-  const { unitsPerEm, ascender, descender } =
-    FONT_METADATA[fontFamily]?.metrics ||
-    FONT_METADATA[FONT_FAMILY.Excalifont].metrics;
+  const { unitsPerEm, ascender, descender } = (
```
Comment (mtolmacs) [inline packages/common/src/font-metadata.ts]: Maybe undefined fontFamily should be reported via `console.error()` so Sentry picks it up. Typescipt shouldn't allow undefined here but since you felt the case should be handled, I assume you find it reasonable that the contract is violated?

Also it would be better to have a fallback font set at a code seam in one place so it can't silently break things if somebody changes the fallback font in some but not all places. The question is whether a broken font rendering is better or a crash.

File: packages/common/src/font-metadata.ts [hunk 2]
```diff
@@ -172,10 +215,40 @@ export const getVerticalOffset = (
 /**
  * Gets line height for a selected family.
  */
-export const getLineHeight = (fontFamily: FontFamilyValues) => {
-  const { lineHeight } =
-    FONT_METADATA[fontFamily]?.metrics ||
-    FONT_METADATA[FONT_FAMILY.Excalifont].metrics;
+export const getLineHeight = (fontFamily: FontFamily) => {
+  const { lineHeight } = (
+    getFontMetadata(fontFamily) ?? FONT_METADATA[FONT_FAMILY.Excalifont]
```
Comment (mtolmacs) [inline packages/common/src/font-metadata.ts]: Same as my previous comment.

File: packages/common/src/font-metadata.ts (hunk 1 above)
Comment (mtolmacs) [reply packages/common/src/font-metadata.ts]: You also seem to have `getFontFamilyFallbacks()` which could be a unification tool if a fallback is indeed needed.
